Optimize zero length input

yinghai opened the pull request:
Summary: Zero length input is something we hit fairly frequently in practice. Previous handling of global TensorPool involves two locks per input (acquire and reclaim). Here we use a specialized anchor tensor to host zero length input. Note that it is only padded to max sequence length. If necessary, an easy extension can be added to pad to max `InputPlaceholder.getType().size()`.

Differential Revision: D19192467

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 37f81aac:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=37f81aac6e5378635ba005d3e37a6d8a1c849616).

### :detective: 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4061527) pytorch_linux_xenial_cuda9_cudnn7_py3_slow_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=78639368) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=78639368))
<details>
<summary>
<code>Dec 25 07:58:13 RuntimeError: test_torch failed!</code>
</summary>

```
Dec 25 07:58:12 Ran 2876 tests in 75.367s 
Dec 25 07:58:12  
Dec 25 07:58:12 FAILED (errors=3, skipped=2862) 
Dec 25 07:58:12  
Dec 25 07:58:12 Generating XML reports... 
Dec 25 07:58:13 Traceback (most recent call last): 
Dec 25 07:58:13   File "test/run_test.py", line 456, in <module> 
Dec 25 07:58:13     main() 
Dec 25 07:58:13   File "test/run_test.py", line 449, in main 
Dec 25 07:58:13     raise RuntimeError(message) 
Dec 25 07:58:13 RuntimeError: test_torch failed! 
Dec 25 07:58:13 + cleanup 
Dec 25 07:58:13 + retcode=1 
Dec 25 07:58:13 + set +x 
Dec 25 07:58:13 =================== sccache compilation log =================== 
Dec 25 07:58:13 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/torch_extensions/test_compilation_error_formatting/main.cpp: In function \'int main()\':\n/tmp/torch_extensions/test_compilation_error_formatting/main.cpp:2:23: error: expected \';\' before \'}\' token\n int main() { return 0 }\n                       ^\n" } 
Dec 25 07:58:13  
Dec 25 07:58:13 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 25 07:58:13 Compile requests               128 
Dec 25 07:58:13 Compile requests executed       47 
Dec 25 07:58:13 Cache hits                      36 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Optimize zero length input (#3943)

Summary:
Pull Request resolved: https://github.com/pytorch/glow/pull/3943

Zero length input is something we hit fairly frequently in practice. Previous handling of global TensorPool involves two locks per input (acquire and reclaim). Here we use a specialized anchor tensor to host zero length input. Note that it is only padded to max sequence length. If necessary, an easy extension can be added to pad to max `InputPlaceholder.getType().size()`.

Reviewed By: jfix71

Differential Revision: D19192467

fbshipit-source-id: bfa7da6bf72f9013001dc7157527fc56c070f913
